### PR TITLE
Feature/d3 asim 2796

### PR DESCRIPTION
**Pull request by hannesdiedrich**

### CODE REVIEW

File: src/d3a/models/area/__init__.py
Comment (spyrostz) [inline src/d3a/models/area/__init__.py:273]: I would remove the load/pv/storage strategy code that calls consume_all_area_commands and call this method only from here. This has the advantage that there is only one place where the aggregator is consuming the commands, instead of calling the same function in the tick method of all the strategies.

File: src/d3a/models/area/redis_external_market_connection.py
Comment (spyrostz) [inline src/d3a/models/area/redis_external_market_connection.py:77]: In order not to forget and to summarize our discussion about it, could you please add a TODO comment in the SimulationConfig class that we should find a more appropriate place for the external_redis_communicator object than the config class.

File: src/d3a/models/area/redis_external_market_connection.py
Comment (spyrostz) [inline src/d3a/models/area/redis_external_market_connection.py:83]: I am curious why this is needed, would be great if you took some time to explain this. I also stumbled upon a similar change in a d3a-web PR, could it be that these 2 are related ?

File: src/d3a/models/area/redis_external_market_connection.py
Comment (spyrostz) [inline src/d3a/models/area/redis_external_market_connection.py:91]: This would not be needed IMO if the transaction_id was popped and not `get` from the payload_data 3 lines above.

File: src/d3a/models/area/redis_external_market_connection.py
Comment (spyrostz) [inline src/d3a/models/area/redis_external_market_connection.py:126]: Same as above

File: src/d3a/models/area/redis_external_market_connection.py
Comment (hannesdiedrich) [reply src/d3a/models/area/redis_external_market_connection.py:77]: Ah, yes, how about in the new added global_objects module in the frame of D3ASIM-2765:
https://github.com/gridsingularity/d3a/pull/879/files#diff-2756a77886549ce43c1ecfd27073b805 ?
Shall I add a story for that? I fear that A TODO will not change much soon ;-)

**Review by spyrostz — DISMISSED** [review]
Some comments, LGTM once resolved/answered

File: src/d3a/models/area/redis_external_market_connection.py
Comment (hannesdiedrich) [reply src/d3a/models/area/redis_external_market_connection.py:83]: IIRC, this is needed for the batch_commands. Here the individual commands are already parsed dicts, but if this method is called directly via the endpoint, the payload["data"] is a string.

File: src/d3a/models/area/redis_external_market_connection.py
Comment (hannesdiedrich) [reply src/d3a/models/area/redis_external_market_connection.py:91]: So you are suggest to rather add it to the retval for the case when the market is not aggregator_controlled and not pop it here?

Comment (spyrostz) [reply src/d3a/models/area/redis_external_market_connection.py:91]: Ah I confused the logic here, so this is the inverse than what I thought, transaction_id is always populated and deleted if we have an aggregator and transaction_ids do not make sense for individual commands. 
I think it would be cleaner though, if the transaction_id was not populated at all in the aggregator case and was only populated in individual commands. This is a matter of personal preference though and YMMV, therefore feel free to ignore this comment and its follow-up. It is fine as it is even now

File: src/d3a/models/area/redis_external_market_connection.py
Comment (spyrostz) [reply src/d3a/models/area/redis_external_market_connection.py:126]: Discussed above, you can ignore it

File: src/d3a/models/area/redis_external_market_connection.py
Comment (spyrostz) [reply src/d3a/models/area/redis_external_market_connection.py:83]: Hmm, I assume you mean this method `trigger_aggregator_commands` when you mention `but if this method is called directly via the endpoint,` . How about doing a json.loads from the trigger_aggregator_commands method, so that the `market_stats_callback` accepts only dicts ?

File: src/d3a/models/area/redis_external_market_connection.py
Comment (hannesdiedrich) [reply src/d3a/models/area/redis_external_market_connection.py:91]: Oh no, you convinced me actually. It would be more clearer your way. Will change this, thanks!

File: src/d3a/models/area/redis_external_market_connection.py
Comment (spyrostz) [reply src/d3a/models/area/redis_external_market_connection.py:77]: Please go ahead to add a story, you are right that TODOs can be long-lived :) thanks!

Comment (hannesdiedrich) [reply src/d3a/models/area/redis_external_market_connection.py:77]: FYI: https://gridsingularity.atlassian.net/browse/D3ASIM-2868

File: src/d3a/models/area/redis_external_market_connection.py
Comment (hannesdiedrich) [reply src/d3a/models/area/redis_external_market_connection.py:83]: I will actually have a look again why this is different for the devices and markets. Will keep you posted

Comment (spyrostz) [reply src/d3a/models/area/redis_external_market_connection.py:83]: Thanks for checking it out!

File: src/d3a/models/area/redis_external_market_connection.py
Comment (faizan2590) [inline src/d3a/models/area/redis_external_market_connection.py:54]: Nice to use predefined utility function `key_in_dict_and_not_none` here.

File: src/d3a/models/area/redis_external_market_connection.py
Comment (hannesdiedrich) [reply src/d3a/models/area/redis_external_market_connection.py:83]: I found out that it is actually the other way around. for the individual commands payload["data"] is a string, but for the batch commands, it is already deserialised to a dict. In the devices, we have different functions for the aggregator and for the direct commands. Here, i propose to keep the if/then implementation and not implement an extra function. Do you agree?

Comment (spyrostz) [reply src/d3a/models/area/redis_external_market_connection.py:83]: Agreed, in this case there is no need to change the implementation

File: src/d3a/models/area/__init__.py
Comment (faizan2590) [inline src/d3a/models/area/__init__.py:118]: Please remove before merging

**Review by faizan2590 — APPROVED** [review]
LGTM

**Pull request merged**
